File: 2023/12/sol.py
import functools

def parse(line):
    s, groups = line.strip().split(" ")
    lookup = {"#": 2, "?": 1, ".": 0}
    return tuple(lookup[char] for char in s), tuple(int(g) for g in groups.split(","))

with open("input") as f:
    data = [parse(x) for x in f]


def match_beginning(data, length):
    return all(x > 0 for x in data[:length]) and (
        (len(data) == length) or data[length] < 2
    )


@functools.cache
def count(data, blocks):
    l2 = {2: '#', 1: '?', 0: '.'}
    total = sum(blocks)
    minimum = sum(x == 2 for x in data)
    maximum = sum(x > 0 for x in data)
    s = ''.join(l2[x] for x in data)
    print("s = {}, c = {}, min = {}, max = {}, total = {}".format(s, blocks, minimum, maximum, total))
    if minimum > total or maximum < total:
        return 0
    if total == 0:
        return 1
    if data[0] == 0:
        return count(data[1:], blocks)
    if data[0] == 2:
        l = blocks[0]
        if match_beginning(data, l):
            if l == len(data):
                return 1
            return count(data[l + 1:], blocks[1:])
        return 0
    return count(data[1:], blocks) + count((2,) + data[1:], blocks)


# counts = [count(*line) for line in data]

# with open("output1", "w") as f:
#     f.write("\n".join(str(x) for x in counts))

# print(sum(count(*line) for line in data))
d = (1,)
b = ()

test, counts = parse("????#.?.?#?? 3,1")
print(count(test, counts))


s = "#.?.?#??"
c = "3,1"
lookup = {"#": 2, "?": 1, ".": 0}
print(match_beginning([lookup[x] for x in s], 3))

File: 2023/12/sol.jl
using Memoization;
operational = '.'
damaged = '#'
unknown = '?'

string_map = Dict('.' => 0, '?' => 1, '#' => 2)

example = """???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1"""

function count_combs(str)
    spring_map, counts = split(str, " ")
    counts = parse.(Int, split(counts, ","))
    return count_string(spring_map, counts)
end


@memoize Dict function count_string(spring_map, counts)
    total = sum(counts)
    if spring_map != ""
        min_value = sum(x -> x == damaged, spring_map)
        max_value = sum(x -> (x in [damaged,unknown]), spring_map)
    else
        min_value = 0
        max_value = 0
    end
    if (min_value > total) || (max_value < total)
        return 0
    end
    if total == 0
        return 1
    end
    if spring_map[1] == operational
        return count_string(spring_map[2:end], counts)
    end
    if spring_map[1] == damaged
        l = counts[1]
        if match_beginning(spring_map, l)
            if l == length(spring_map)
                return 1
            end
            return count_string(spring_map[l+2:end], counts[2:end])
        end
        return 0
    end
    return count_string(spring_map[2:end], counts) + count_string('#' * spring_map[2:end], counts)
end

function match_beginning(data, l)
    c1 = all(map(x->(x in [damaged, unknown]), collect(data[1:l])))
    c2 = (length(data) == l) || (data[l+1] .== operational) || (data[l+1] .== unknown)
    return c1 && c2
end

# count_string(test_str, counts) == 1

function sol1(input)
    total = 0
    for (i,line) in enumerate(split(input, "\n"))
        c = count_combs(line)
        total += c
    end
    return total
end

function dupe_line(line)
    spring_map, counts = split(line, " ")
    spring_map = join([spring_map, spring_map, spring_map, spring_map, spring_map], "?")
    counts = join([counts, counts, counts, counts, counts], ",")
    return join([spring_map, counts], " ")
end

function sol2(input)
    total = 0
    for (i,line) in enumerate(split(input, "\n"))
        d1 = dupe_line(line)
        c = count_combs(d1)
        total += c
    end
    return total
end

input = read("2023/12/input", String)

sol1(input)

@time sol2(input)



